src/hooks/useQueryParams.tsx

```diff
@@ -0,0 +1,60 @@
+import { isEqual } from "lodash";
+import queryString from "query-string";
+import { Dispatch, SetStateAction, useEffect, useState } from "react";
+import { useHistory, useLocation } from "react-router-dom";
+
+export interface QueryParams {
+  [key: string]: string | number | undefined;
+}
+
+export type HistoryAction = "replace" | "push";
+
+// A basic hook for fetching query param values and setting query param values
+const useQueryParams = (
+  defaultValues: QueryParams,
+  defaultAction: HistoryAction = "replace",
+): [
+  QueryParams,
+  Dispatch<SetStateAction<QueryParams>>,
+  Dispatch<SetStateAction<HistoryAction>>,
+] => {
+  const history = useHistory();
+  const location = useLocation();
+
+  const defaultQueryParams = queryString.parse(location.search) as QueryParams;
+  const hasExistingURLQueryParams = Object.keys(defaultQueryParams).length > 0;
```

Comment:
Note: Sometimes a custom query param will get added without the user's consent. For example, Facebook adds `fbclid` or something like that to links you click that take you away from Facebook. We'll probably need to let the hook user pass in a custom way to detect whether the existing query params actually count as "has existing".